packages/lesswrong/server/editor/conversionUtils.ts

```diff
@@ -412,3 +413,160 @@ export async function dataToWordCount(data: AnyBecauseTodo, type: string) {
+    newFootnoteContent.append(footnoteContent.contents());
+
+    const newFootnoteItem = $('<li class="footnote-item" data-footnote-item="" data-footnote-id="' + id + '" data-footnote-index="' + index + '" role="doc-endnote" id="fn' + id + '"></li>');
+    newFootnoteItem.append(newFootnoteBackLink);
+    newFootnoteItem.append(newFootnoteContent);
+
+    $('.footnote-section').append(newFootnoteItem);
+    item.remove()
+  });
+
+  // The changes so far leave over a stub like so:
+  // <hr />
+  //   <div>
+  //     <p></p>
+  //   </div>
+  //   ...
+  // <ol class=\"footnotes\" role=\"doc-endnotes\">
+  //
+  // Remove everything from the <hr /> to the footnotes section
+  const footnotesSection = $('.footnote-section');
+  const hrBeforeFootnotes = footnotesSection.prevAll('hr').last();
+  hrBeforeFootnotes.remove();
+
+  return $.html();
+}
+
+/**
+ * Remote the google redirect from links that come from google docs
+ *
+ * https://www.google.com/url?q=https://en.wikipedia.org/wiki/Main_Page becomes https://en.wikipedia.org/wiki/Main_Page
+ */
+function googleDocRemoveRedirects(html: string): string {
+  const $ = cheerio.load(html);
+
+  // Regex match examples:
+  // https://www.google.com/url?q=https://en.wikipedia.org/wiki/Main_Page&sa=D&source=editors&ust=1667922372715536&usg=AOvVaw2NyT5CZhfsrRY_zzMs2UUJ
+  //                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ <- first match group matches this, stopping at the first &
+  // https://www.google.com/url?q=https://en.wikipedia.org/wiki/Main_Page
+  //                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ <- if there are no more params (no &), match up to the end of the string
+  const hrefPattern = /^https:\/\/www\.google\.com\/url\?q=(\S+?)(&|$)/;
+
+  $('a[href]').each((_, element) => {
+    const href = $(element).attr('href');
+    if (!href) return
+
+    const match = hrefPattern.exec(href);
+    if (match && match[1]) {
+      $(element).attr('href', decodeURIComponent(match[1]));
+    }
+  });
+
+  return $.html();
+}
+
+/**
+ * We need to convert a few things in the raw html exported from google to make it work with ckeditor, this is
+ * largely mirroring conversions we do on paste in the ckeditor code:
+ * - Convert footnotes to our format
+ * - Remove google redirects from all urls
+ * - Reupload images to cloudinary (we actually don't do this on paste, but it's easier to do so here and prevents images breaking due to rate limits)
+ */
+export async function convertImportedGoogleDoc(html: string, postId: string) {
+  const { html: withRehostedImages } = await convertImagesInHTML(html, postId, url => url.includes("googleusercontent"))
+  const withConvertedFootnotes = googleDocConvertFootnotes(withRehostedImages)
+  const withNormalisedRedirects = googleDocRemoveRedirects(withConvertedFootnotes)
+  const ckEditorMarkup = await dataToCkEditor(withNormalisedRedirects, "html")
```

Comment:
A couple of import issues I noticed that probably need to be handled here:

- The document title got imported as normal text at the top of the document, but maybe we should use it to set the post title and remove it from the document body?
- Bold, italics and underline don't work
- Internal links still point to the google doc, maybe we can fix this now or is it still difficult?

Reply:
Ok I did the bold/italics one, and made a [ticket](https://app.asana.com/0/0/1206726735026907/f) for internal links